In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#Scaling
from sklearn.preprocessing import MinMaxScaler

 #Correlation Heatmap
from matplotlib.colors import LinearSegmentedColormap

#Statistical Test
from scipy import stats
from sklearn.impute import SimpleImputer

#Models
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split

from sklearn.model_selection import KFold, RepeatedKFold, cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression, RidgeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report
from sklearn.model_selection import train_test_split, cross_val_score

from sklearn.linear_model import Ridge

pd.set_option('display.max_columns', None)

from sklearn.metrics import f1_score
import time
from sklearn.model_selection import StratifiedKFold

In [2]:
# You can download the data in the source that is linked above the table of contents

# Read in the data
X_val_encoded = pd.read_csv('../project_data/X_val_encoded.csv', delimiter=',', index_col=0)
X_train_encoded = pd.read_csv('../project_data/X_train_encoded.csv', delimiter=',', index_col=0)

y_train = pd.read_csv('../project_data/y_train.csv',delimiter=',', index_col=0)
y_val= pd.read_csv('../project_data/y_val.csv', delimiter=',', index_col=0)

X_test_encoded = pd.read_csv('../project_data/X_test_encoded.csv',index_col=0)

In [3]:
#NEW: just to check the shape
print(X_train_encoded.shape[0])
print(y_train.shape[0])

382537
382537


In [4]:
claim_injury_type_mapping = {
    '4. TEMPORARY': 4,
    '2. NON-COMP': 2,
    '5. PPD SCH LOSS': 5,
    '3. MED ONLY': 3,
    '6. PPD NSL': 6,
    '1. CANCELLED': 1,
    '8. DEATH': 8,
    '7. PTD': 7
}

y_train_encoded = y_train['Claim Injury Type'].map(claim_injury_type_mapping)
y_val_encoded = y_val['Claim Injury Type'].map(claim_injury_type_mapping)
#NEW
y_train_encoded = pd.DataFrame(y_train_encoded, columns=['Claim Injury Type'])
y_val_encoded = pd.DataFrame(y_val_encoded, columns=['Claim Injury Type'])

In [5]:
num_columns = ['Age at Injury', 'Average Weekly Wage', 'Birth Year', 'IME-4 Count', 'Number of Dependents', 'Days_between_Acc_Assembyl']

cat_columns = ['Alternative Dispute Resolution', 'Attorney/Representative', 'C-2 Date', 'C-3 Date', 'COVID-19 Indicator', 'Gender','First Hearing Date',
               'CarrierGroup_Self-insured Private Entity', 'CarrierGroup_Self-insured Public Entity', 'CarrierGroup_Special Funds',
               'CarrierGroup_State Insurance Fund', 'Industry Code_encoded_5. PPD SCH LOSS', 'Industry Code_encoded_2. NON-COMP', 
               'Industry Code_encoded_3. MED ONLY', 'Industry Code_encoded_4. TEMPORARY', 'Industry Code_encoded_1. CANCELLED', 
               'Industry Code_encoded_8. DEATH', 'Industry Code_encoded_6. PPD NSL', 'Industry Code_encoded_7. PTD',
               'Injury_Cause_Category_encoded_5. PPD SCH LOSS', 'Injury_Cause_Category_encoded_2. NON-COMP', 
               'Injury_Cause_Category_encoded_3. MED ONLY', 'Injury_Cause_Category_encoded_4. TEMPORARY', 
               'Injury_Cause_Category_encoded_1. CANCELLED', 'Injury_Cause_Category_encoded_8. DEATH','Injury_Cause_Category_encoded_6. PPD NSL',
                'Injury_Cause_Category_encoded_7. PTD', 'Nature_Injury_Occupational', 'Nature_Injury_Specific', 'Nature_Injury_Unknown',
                'Part_Body_Category_encoded_5. PPD SCH LOSS', 'Part_Body_Category_encoded_2. NON-COMP', 'Part_Body_Category_encoded_3. MED ONLY',
                'Part_Body_Category_encoded_4. TEMPORARY', 'Part_Body_Category_encoded_1. CANCELLED', 'Part_Body_Category_encoded_8. DEATH',
                'Part_Body_Category_encoded_6. PPD NSL', 'Part_Body_Category_encoded_7. PTD']


# Create subsets
X_train_num = X_train_encoded[num_columns]
X_train_cat = X_train_encoded[cat_columns]

X_val_num = X_val_encoded[num_columns]
X_val_cat = X_val_encoded[cat_columns]

X_test_num=X_test_encoded[num_columns]
X_test_cat=X_test_encoded[cat_columns]

In [6]:
scaler = MinMaxScaler().fit(X_train_num)
X_train_num_scaled = scaler.transform(X_train_num)

# Convert the array to a pandas dataframe
X_train_num_scaled = pd.DataFrame(X_train_num_scaled, columns = X_train_num.columns).set_index(X_train_encoded.index)
X_val_num_scaled = scaler.transform(X_val_num)

X_test_num_scaled = scaler.transform(X_test_num)

In [7]:
# Remerge the numerical and categorical features
X_train = pd.concat([X_train_num[['IME-4 Count']], X_train_cat], axis=1)
X_val = pd.concat([X_val_num[['IME-4 Count']], X_val_cat], axis=1)
X_test =pd.concat([X_test_num[['IME-4 Count']], X_test_cat], axis=1)
y_train=y_train_encoded.copy()
y_val=y_val_encoded.copy()

In [8]:
#transform into a series
y_train = y_train.squeeze()
y_val = y_val.squeeze()

In [10]:
!pip install lightgbm

Defaulting to user installation because normal site-packages is not writeable


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip



   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   --------------------- ------------------ 0.8/1.4 MB 4.2 MB/s eta 0:00:01
   ---------------------------------------- 1.4/1.4 MB 4.0 MB/s eta 0:00:00


---

In [ ]:
from sklearn.linear_model import LogisticRegression, RidgeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.metrics import classification_report, confusion_matrix, f1_score

# Function to calculate and print metrics
def metrics(y_train, pred_train, y_val, pred_val):
    print('___________________________________________________________________________________________________________')
    print('                                                     TRAIN                                                 ')
    print('-----------------------------------------------------------------------------------------------------------')
    print(classification_report(y_train, pred_train))
    print('Confusion Matrix (Train):\n', confusion_matrix(y_train, pred_train))
    print('\nF1 Macro Score (Train):', f1_score(y_train, pred_train, average='macro'))

    print('___________________________________________________________________________________________________________')
    print('                                                VALIDATION                                                 ')
    print('-----------------------------------------------------------------------------------------------------------')
    print(classification_report(y_val, pred_val))
    print('Confusion Matrix (Validation):\n', confusion_matrix(y_val, pred_val))
    print('\nF1 Macro Score (Validation):', f1_score(y_val, pred_val, average='macro'))

# List of models to evaluate
models = {
    "Logistic Regression": LogisticRegression(max_iter=500, random_state=42),
    "Neural Network": MLPClassifier(max_iter=500, random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "SVM (RBF Kernel)": SVC(kernel='rbf', probability=True, random_state=42),
    #"Gradient Boosting": GradientBoostingClassifier(random_state=42),
    #"XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42),
    "LightGBM": LGBMClassifier(random_state=42),
    "Ridge Classifier": RidgeClassifier()
}

# Train and evaluate each model
for model_name, model in models.items():
    print(f"\n{'='*50}\nTraining and Evaluating: {model_name}\n{'='*50}")
    
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    pred_train = model.predict(X_train)
    pred_val = model.predict(X_val)
    
    # Print metrics
    metrics(y_train, pred_train, y_val, pred_val)


Training and Evaluating: Logistic Regression


C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


___________________________________________________________________________________________________________
                                                     TRAIN                                                 
-----------------------------------------------------------------------------------------------------------


C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.

              precision    recall  f1-score   support

           1       0.67      0.43      0.52      7701
           2       0.71      0.94      0.81    200073
           3       0.30      0.01      0.01     46185
           4       0.57      0.53      0.55     95874
           5       0.53      0.40      0.45     29651
           6       0.00      0.00      0.00      2721
           7       0.00      0.00      0.00        57
           8       0.27      0.06      0.10       275

    accuracy                           0.67    382537
   macro avg       0.38      0.30      0.31    382537
weighted avg       0.61      0.67      0.61    382537

Confusion Matrix (Train):
 [[  3307   3324     26    815    226      0      0      3]
 [  1499 188290    241   7825   2205      0      0     13]
 [    50  31188    354  11884   2706      0      0      3]
 [    44  39143    448  50937   5274      0      0     28]
 [    11   2545     92  15175  11828      0      0      0]
 [     0     12     22   24

C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.

              precision    recall  f1-score   support

           1       0.68      0.44      0.53      3263
           2       0.71      0.94      0.81     85765
           3       0.29      0.01      0.01     19810
           4       0.57      0.53      0.55     41064
           5       0.53      0.39      0.45     12669
           6       0.00      0.00      0.00      1167
           7       0.00      0.00      0.00        24
           8       0.38      0.09      0.14       123

    accuracy                           0.67    163885
   macro avg       0.39      0.30      0.31    163885
weighted avg       0.60      0.67      0.61    163885

Confusion Matrix (Validation):
 [[ 1431  1368     4   366    94     0     0     0]
 [  633 80694   100  3350   985     0     0     3]
 [   26 13311   150  5154  1165     0     0     4]
 [   19 16778   215 21885  2156     0     0    11]
 [    4  1106    47  6529  4983     0     0     0]
 [    0     6     3  1081    77     0     0     0]
 [    0    

C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.

              precision    recall  f1-score   support

           1       0.69      0.48      0.56      7701
           2       0.72      0.94      0.82    200073
           3       0.47      0.02      0.04     46185
           4       0.58      0.57      0.58     95874
           5       0.61      0.43      0.50     29651
           6       0.00      0.00      0.00      2721
           7       0.00      0.00      0.00        57
           8       0.43      0.26      0.33       275

    accuracy                           0.68    382537
   macro avg       0.44      0.34      0.35    382537
weighted avg       0.64      0.68      0.63    382537

Confusion Matrix (Train):
 [[  3673   3038     25    861     98      0      0      6]
 [  1493 188824    330   8161   1234      0      0     31]
 [    51  30476    917  12509   2220      0      0     12]
 [    85  36296    581  54424   4443      0      0     45]
 [    35   2349    103  14435  12729      0      0      0]
 [     0     20      3   25

C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.

              precision    recall  f1-score   support

           1       0.69      0.48      0.56      3263
           2       0.72      0.94      0.82     85765
           3       0.44      0.02      0.04     19810
           4       0.58      0.57      0.58     41064
           5       0.61      0.42      0.50     12669
           6       0.00      0.00      0.00      1167
           7       0.00      0.00      0.00        24
           8       0.41      0.23      0.29       123

    accuracy                           0.68    163885
   macro avg       0.43      0.33      0.35    163885
weighted avg       0.64      0.68      0.63    163885

Confusion Matrix (Validation):
 [[ 1558  1264    12   383    42     0     0     4]
 [  622 80896   155  3518   564     0     0    10]
 [   32 13058   379  5384   951     0     0     6]
 [   37 15495   272 23377  1863     0     0    20]
 [   19  1017    36  6234  5363     0     0     0]
 [    0     5     3  1099    60     0     0     0]
 [    0    

==================================================
Training and Evaluating: Logistic Regression
==================================================
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
___________________________________________________________________________________________________________
                                                     TRAIN                                                 
-----------------------------------------------------------------------------------------------------------
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
              precision    recall  f1-score   support

           1       0.67      0.43      0.52      7701
           2       0.71      0.94      0.81    200073
           3       0.30      0.01      0.01     46185
           4       0.57      0.53      0.55     95874
           5       0.53      0.40      0.45     29651
           6       0.00      0.00      0.00      2721
           7       0.00      0.00      0.00        57
           8       0.27      0.06      0.10       275

    accuracy                           0.67    382537
   macro avg       0.38      0.30      0.31    382537
weighted avg       0.61      0.67      0.61    382537

Confusion Matrix (Train):
 [[  3307   3324     26    815    226      0      0      3]
 [  1499 188290    241   7825   2205      0      0     13]
 [    50  31188    354  11884   2706      0      0      3]
 [    44  39143    448  50937   5274      0      0     28]
 [    11   2545     92  15175  11828      0      0      0]
 [     0     12     22   2462    225      0      0      0]
 [     1      0      0     54      2      0      0      0]
 [     3     69      8    175      3      0      0     17]]

F1 Macro Score (Train): 0.3067390527176675
___________________________________________________________________________________________________________
                                                VALIDATION                                                 
-----------------------------------------------------------------------------------------------------------
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
              precision    recall  f1-score   support

           1       0.68      0.44      0.53      3263
           2       0.71      0.94      0.81     85765
           3       0.29      0.01      0.01     19810
           4       0.57      0.53      0.55     41064
           5       0.53      0.39      0.45     12669
           6       0.00      0.00      0.00      1167
           7       0.00      0.00      0.00        24
           8       0.38      0.09      0.14       123

    accuracy                           0.67    163885
   macro avg       0.39      0.30      0.31    163885
weighted avg       0.60      0.67      0.61    163885

Confusion Matrix (Validation):
 [[ 1431  1368     4   366    94     0     0     0]
 [  633 80694   100  3350   985     0     0     3]
 [   26 13311   150  5154  1165     0     0     4]
 [   19 16778   215 21885  2156     0     0    11]
 [    4  1106    47  6529  4983     0     0     0]
 [    0     6     3  1081    77     0     0     0]
 [    0     0     0    23     1     0     0     0]
 [    2    27     4    78     1     0     0    11]]

F1 Macro Score (Validation): 0.3128883850064663

==================================================
Training and Evaluating: Neural Network
==================================================
___________________________________________________________________________________________________________
                                                     TRAIN                                                 
-----------------------------------------------------------------------------------------------------------
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
              precision    recall  f1-score   support

           1       0.69      0.48      0.56      7701
           2       0.72      0.94      0.82    200073
           3       0.47      0.02      0.04     46185
           4       0.58      0.57      0.58     95874
           5       0.61      0.43      0.50     29651
           6       0.00      0.00      0.00      2721
           7       0.00      0.00      0.00        57
           8       0.43      0.26      0.33       275

    accuracy                           0.68    382537
   macro avg       0.44      0.34      0.35    382537
weighted avg       0.64      0.68      0.63    382537

Confusion Matrix (Train):
 [[  3673   3038     25    861     98      0      0      6]
 [  1493 188824    330   8161   1234      0      0     31]
 [    51  30476    917  12509   2220      0      0     12]
 [    85  36296    581  54424   4443      0      0     45]
 [    35   2349    103  14435  12729      0      0      0]
 [     0     20      3   2517    181      0      0      0]
 [     0      0      0     55      2      0      0      0]
 [     4     45      4    150      0      0      0     72]]

F1 Macro Score (Train): 0.3532944078089757
___________________________________________________________________________________________________________
                                                VALIDATION                                                 
-----------------------------------------------------------------------------------------------------------
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
              precision    recall  f1-score   support

           1       0.69      0.48      0.56      3263
           2       0.72      0.94      0.82     85765
           3       0.44      0.02      0.04     19810
           4       0.58      0.57      0.58     41064
           5       0.61      0.42      0.50     12669
           6       0.00      0.00      0.00      1167
           7       0.00      0.00      0.00        24
           8       0.41      0.23      0.29       123

    accuracy                           0.68    163885
   macro avg       0.43      0.33      0.35    163885
weighted avg       0.64      0.68      0.63    163885

Confusion Matrix (Validation):
 [[ 1558  1264    12   383    42     0     0     4]
 [  622 80896   155  3518   564     0     0    10]
 [   32 13058   379  5384   951     0     0     6]
 [   37 15495   272 23377  1863     0     0    20]
 [   19  1017    36  6234  5363     0     0     0]
 [    0     5     3  1099    60     0     0     0]
 [    0     0     0    23     1     0     0     0]
 [    3    25     0    66     1     0     0    28]]

F1 Macro Score (Validation): 0.3483434719716329

==================================================
Training and Evaluating: Random Forest
==================================================
___________________________________________________________________________________________________________
                                                     TRAIN                                                 
-----------------------------------------------------------------------------------------------------------
              precision    recall  f1-score   support

           1       0.75      0.54      0.63      7701
           2       0.75      0.95      0.84    200073
           3       0.66      0.19      0.30     46185
           4       0.72      0.60      0.65     95874
           5       0.70      0.67      0.68     29651
           6       0.84      0.31      0.45      2721
           7       0.89      0.56      0.69        57
           8       0.73      0.61      0.67       275

    accuracy                           0.74    382537
   macro avg       0.75      0.55      0.61    382537
weighted avg       0.73      0.74      0.71    382537

Confusion Matrix (Train):
 [[  4148   2764    161    514    112      0      0      2]
 [  1180 190307   1529   5865   1176      3      1     12]
 [    71  27259   8967   7558   2290     28      0     12]
 [   119  31615   2109  57058   4861     80      2     30]
 [    28   1888    788   6925  19975     44      0      3]
 [     0     26     64   1514    281    833      1      2]
 [     0      2      1     20      2      0     32      0]
 [     0     30     23     52      1      0      0    169]]

F1 Macro Score (Train): 0.6133063835157717
___________________________________________________________________________________________________________
                                                VALIDATION                                                 
-----------------------------------------------------------------------------------------------------------
              precision    recall  f1-score   support

           1       0.63      0.46      0.53      3263
           2       0.73      0.92      0.81     85765
           3       0.26      0.07      0.11     19810
           4       0.58      0.49      0.53     41064
           5       0.51      0.49      0.50     12669
           6       0.07      0.02      0.03      1167
           7       0.00      0.00      0.00        24
           8       0.20      0.11      0.15       123

    accuracy                           0.66    163885
   macro avg       0.37      0.32      0.33    163885
weighted avg       0.61      0.66      0.62    163885

Confusion Matrix (Validation):
 [[ 1494  1278    85   323    81     0     0     2]
 [  629 79248  1227  3893   749     6     0    13]
 [   73 12349  1448  4601  1294    30     2    13]
 [  124 14801  2031 20264  3606   201    11    26]
 [   31  1053   634  4703  6211    37     0     0]
 [    1     8    45   952   140    21     0     0]
 [    0     0     0    20     2     1     0     1]
 [    2    36    12    58     0     1     0    14]]

F1 Macro Score (Validation): 0.3339612886975133

==================================================
Training and Evaluating: Decision Tree
==================================================
___________________________________________________________________________________________________________
                                                     TRAIN                                                 
-----------------------------------------------------------------------------------------------------------
              precision    recall  f1-score   support

           1       0.71      0.58      0.64      7701
           2       0.74      0.96      0.84    200073
           3       0.63      0.21      0.32     46185
           4       0.73      0.58      0.65     95874
           5       0.73      0.62      0.67     29651
           6       0.95      0.27      0.42      2721
           7       0.97      0.51      0.67        57
           8       0.87      0.51      0.65       275

    accuracy                           0.74    382537
   macro avg       0.79      0.53      0.61    382537
weighted avg       0.73      0.74      0.71    382537

Confusion Matrix (Train):
 [[  4466   2617    119    415     84      0      0      0]
 [  1369 191881   1070   4831    919      0      0      3]
 [   153  28030   9767   6357   1870      5      0      3]
 [   228  32831   3105  55964   3706     26      1     13]
 [    61   2188   1179   7715  18501      6      0      1]
 [     2     39    114   1555    269    741      0      1]
 [     0      4      0     21      2      1     29      0]
 [     2     43     32     57      0      0      0    141]]

F1 Macro Score (Train): 0.606305457131044
___________________________________________________________________________________________________________
                                                VALIDATION                                                 
-----------------------------------------------------------------------------------------------------------
              precision    recall  f1-score   support

           1       0.56      0.48      0.51      3263
           2       0.72      0.93      0.81     85765
           3       0.24      0.08      0.12     19810
           4       0.59      0.46      0.52     41064
           5       0.52      0.46      0.49     12669
           6       0.10      0.03      0.05      1167
           7       0.00      0.00      0.00        24
           8       0.18      0.11      0.14       123

    accuracy                           0.66    163885
   macro avg       0.36      0.32      0.33    163885
weighted avg       0.61      0.66      0.62    163885

Confusion Matrix (Validation):
 [[ 1550  1244   103   289    74     1     0     2]
 [  779 79557  1297  3427   674     8     3    20]
 [  136 12641  1624  4158  1206    28     2    15]
 [  235 15490  2842 18923  3248   278    22    26]
 [   64  1250   892  4595  5821    45     1     1]
 [    6    15    76   895   135    40     0     0]
 [    0     2     2    16     2     2     0     0]
 [    5    40    15    43     2     3     1    14]]

F1 Macro Score (Validation): 0.33016134284356835

==================================================
Training and Evaluating: SVM (RBF Kernel)
==================================================
Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip

Collecting lightgbm
  Downloading lightgbm-4.5.0-py3-none-win_amd64.whl.metadata (17 kB)
Requirement already satisfied: numpy>=1.17.0 in c:\users\utilizador\appdata\roaming\python\python312\site-packages (from lightgbm) (2.0.2)
Requirement already satisfied: scipy in c:\users\utilizador\appdata\roaming\python\python312\site-packages (from lightgbm) (1.14.1)
Downloading lightgbm-4.5.0-py3-none-win_amd64.whl (1.4 MB)
   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   --------------------- ------------------ 0.8/1.4 MB 4.2 MB/s eta 0:00:01
   ---------------------------------------- 1.4/1.4 MB 4.0 MB/s eta 0:00:00
Installing collected packages: lightgbm
Successfully installed lightgbm-4.5.0

In [ ]:
# from sklearn.model_selection import KFold, RepeatedKFold, StratifiedKFold, cross_val_score
# from sklearn.neural_network import MLPClassifier
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.linear_model import LogisticRegression
# from sklearn.neighbors import KNeighborsClassifier
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.naive_bayes import GaussianNB
# from sklearn.linear_model import RidgeClassifier
# from sklearn.metrics import make_scorer, f1_score

# # Define K-Fold cross-validation
# kfold = KFold(n_splits=5, shuffle=True, random_state=1)

# # Define Repeated K-Fold cross-validation
# repeated_kfold = RepeatedKFold(n_splits=5, n_repeats=3, random_state=1)

# # Define Stratified K-Fold cross-validation
# skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=1)

# # Define the F1 macro scorer
# f1_macro_scorer = make_scorer(f1_score, average='macro')

# # Decide the best model
# def evaluate_model(model, X, y, kfold, repeated_kfold, skfold):
#     kfold_results = cross_val_score(model, X, y, cv=kfold, scoring=f1_macro_scorer)
#     repeated_kfold_results = cross_val_score(model, X, y, cv=repeated_kfold, scoring=f1_macro_scorer)
#     skfold_results = cross_val_score(model, X, y, cv=skfold, scoring=f1_macro_scorer)

#     print(f"Model: {model.__class__.__name__}")
#     print(f"K-Fold F1 Score (Macro) Mean: {kfold_results.mean():.4f} (+/- {kfold_results.std():.4f})")
#     print(f"Repeated K-Fold F1 Score (Macro) Mean: {repeated_kfold_results.mean():.4f} (+/- {repeated_kfold_results.std():.4f})")
#     print(f"Stratified K-Fold F1 Score (Macro) Mean: {skfold_results.mean():.4f} (+/- {skfold_results.std():.4f})")
#     print("")

# # Define models
# models = {
#     "MLPClassifier": MLPClassifier(max_iter=500),
#     "RandomForest": RandomForestClassifier(),
#     "LogisticRegression": LogisticRegression(max_iter=1000),
#     "KNeighborsClassifier": KNeighborsClassifier(),
#     "DecisionTree": DecisionTreeClassifier(),
#     "GaussianNB": GaussianNB(),
#     "RidgeClassifier": RidgeClassifier()
# }

# # Evaluate each model using K-Fold and Repeated K-Fold cross-validation
# for model_name, model in models.items():
#     print(f"Evaluating {model_name}...")
#     evaluate_model(model, X_train, y_train, kfold, repeated_kfold, skf)

Evaluating MLPClassifier...
Model: MLPClassifier
K-Fold F1 Score (Macro) Mean: 0.3361 (+/- 0.0074)
Repeated K-Fold F1 Score (Macro) Mean: 0.3413 (+/- 0.0074)
Stratified K-Fold F1 Score (Macro) Mean: 0.3379 (+/- 0.0117)

Evaluating RandomForest...
Model: RandomForestClassifier
K-Fold F1 Score (Macro) Mean: 0.3382 (+/- 0.0081)
Repeated K-Fold F1 Score (Macro) Mean: 0.3363 (+/- 0.0082)
Stratified K-Fold F1 Score (Macro) Mean: 0.3358 (+/- 0.0117)

Evaluating LogisticRegression...


C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Utilizador\AppData\Roaming\Python\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules

Model: LogisticRegression
K-Fold F1 Score (Macro) Mean: 0.2996 (+/- 0.0034)
Repeated K-Fold F1 Score (Macro) Mean: 0.3014 (+/- 0.0042)
Stratified K-Fold F1 Score (Macro) Mean: 0.3020 (+/- 0.0059)

Evaluating KNeighborsClassifier...
Model: KNeighborsClassifier
K-Fold F1 Score (Macro) Mean: 0.3191 (+/- 0.0031)
Repeated K-Fold F1 Score (Macro) Mean: 0.3198 (+/- 0.0038)
Stratified K-Fold F1 Score (Macro) Mean: 0.3185 (+/- 0.0087)

Evaluating DecisionTree...
Model: DecisionTreeClassifier
K-Fold F1 Score (Macro) Mean: 0.3342 (+/- 0.0070)
Repeated K-Fold F1 Score (Macro) Mean: 0.3263 (+/- 0.0086)
Stratified K-Fold F1 Score (Macro) Mean: 0.3265 (+/- 0.0140)

Evaluating GaussianNB...
Model: GaussianNB
K-Fold F1 Score (Macro) Mean: 0.1591 (+/- 0.0025)
Repeated K-Fold F1 Score (Macro) Mean: 0.1592 (+/- 0.0029)
Stratified K-Fold F1 Score (Macro) Mean: 0.1597 (+/- 0.0033)

Evaluating RidgeClassifier...
Model: RidgeClassifier
K-Fold F1 Score (Macro) Mean: 0.2559 (+/- 0.0012)
Repeated K-Fold F1 Score

# Define models
models = {
    'GaussianNB': GaussianNB(),
    'KNN': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'SVM (balanced)': SVC(class_weight='balanced'),
    "SVM": SVC(),
    'AdaBoost': AdaBoostClassifier(),
    'Random Forest (balanced)': RandomForestClassifier(class_weight='balanced'),
    'Random Forest': RandomForestClassifier(),
    'Neural Network': MLPClassifier(solver='adam',max_iter=1000),
}

# F1_Scores
f1_scores = []

for model_name, model in models.items():
    print(f"Fitting {model_name}...")
    
    # Use the model directly on the validation set
    y_pred_val = model.fit(X_mult_train, y_mult_train).predict(X_mult_val)
    f1_val = f1_score(y_mult_val, y_pred_val, pos_label=1,average="weighted")
    
    f1_scores.append([f1_val])

# Create DataFrame with results
f1_res = pd.DataFrame({
    "F1_Scores": [result[0] for result in f1_scores],
    "Algorithm": list(models.keys())
})

from sklearn.metrics import classification_report, confusion_matrix

def metrics(y_train, pred_train , y_val, pred_val):
    print('___________________________________________________________________________________________________________')
    print('                                                     TRAIN                                                 ')
    print('-----------------------------------------------------------------------------------------------------------')
    print(classification_report(y_train, pred_train))
    print(confusion_matrix(y_train, pred_train))


    print('___________________________________________________________________________________________________________')
    print('                                                VALIDATION                                                 ')
    print('-----------------------------------------------------------------------------------------------------------')
    print(classification_report(y_val, pred_val))
    print(confusion_matrix(y_val, pred_val))

Modelos Tradicionais de Machine Learning

Logistic Regression Multiclasse
Funcionamento: Extensão da regressão logística binária para prever várias categorias com técnicas como One-vs-Rest (OvR) ou Softmax.
Vantagens: Simples de entender, eficiente em datasets pequenos.
Limitações: Não lida bem com relações complexas entre as variáveis ou entre os rótulos.
Escalabilidade: Precisa de dados normalizados.

Decision Trees (DT)
Funcionamento: Constrói árvores de decisão com base em critérios de impureza (como Gini ou Entropia).
Vantagens: Funciona com Ordinal Encoding, não exige normalização.
Limitações: Tende a sobreajustar em datasets pequenos.

Random Forest (RF)
Funcionamento: Ensemble de árvores de decisão para reduzir o sobreajuste.
Vantagens: Robusto contra overfitting, lida bem com categorical encoding.
Limitações: Pode ser mais lento em grandes datasets.

Gradient Boosting (e.g., XGBoost, LightGBM)
Funcionamento: Constrói modelos iterativamente, corrigindo erros das previsões anteriores.
Vantagens: Excelente para dados tabulares com muitas categorias.
Limitações: Exige Ordinal Encoding ou One-Hot Encoding.



Modelos Baseados em Redes Neurais

Multilayer Perceptron (MLP)
Funcionamento: Redes totalmente conectadas com ativação Softmax para saída multicategoria.
Vantagens: Flexível, pode capturar relações complexas.
Limitações: Exige grande volume de dados, sensível à escala.

Convolutional Neural Networks (CNNs)
Funcionamento: Modelos baseados em convolução, ideais para imagens ou séries temporais.
Vantagens: Excelente para dados espaciais.
Limitações: Ineficiente para dados tabulares ou não estruturados.

Recurrent Neural Networks (RNNs) e Transformers
Funcionamento: Ideais para dados sequenciais (texto ou séries temporais).
Vantagens: Capturam dependências temporais.
Limitações: Mais complexos de treinar.
Modelos de Similaridade e Vizinho Mais Próximo
K-Nearest Neighbors (KNN)
Funcionamento: Classifica com base nos rótulos dos vizinhos mais próximos.
Vantagens: Simples de implementar, não precisa de treinamento explícito.
Limitações: Sensível à escala (precisa de normalização).



Modelos Estatísticos

Naive Bayes
Funcionamento: Calcula probabilidades assumindo independência entre atributos.
Vantagens: Rápido, eficiente com dados textuais (TF-IDF).
Limitações: Assume independência entre variáveis (o que nem sempre é verdade).
Considerações sobre Pré-processamento
Encoding de Variáveis Categóricas:
Ordinal Encoding: Adequado para modelos baseados em árvores, mas não para métodos baseados em distância como KNN.
One-Hot Encoding: Necessário para redes neurais e modelos lineares (Logistic Regression, SVM).
Normalização:
Essencial para algoritmos baseados em distância (KNN, SVM) e redes neurais.
Dados Desequilibrados:
Métodos como Oversampling (SMOTE) ou ajuste de pesos nas classes podem ser usados.



!!!!!!!!!!!!!!!!!6 - If you use GridSearch or something similar like RandomSearch please create your final model in another cell that considers the best parameters.